[workflow] Use internal storage

suquark opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This PR replaces the workflow storage with the Internal Ray storage.

Two things remains for later PRs to make this PR easier for review.
* How to use storage with `ray start`? Currently `ray start --storage --head` + `ray.init(address="auto")` => `storage = None`
* Enable debug storage in workflow.

Related tests are disabled temporarily.

Some behavior changes:

Before we have to ensure users are calling `workflow.init` manually before running workflow to setup the storage; and `ray.init` is called automatically if Ray has not been initialized. Now since the storage is setup in `ray.init`, we cannot call `ray.init` automatically, because you have to setup a storage for the workflow. Instead, `workflow.init` is called automatically (and we may consider remove it from API later), and if `workflow.init` detects Ray has not been initialized, it calls `ray.init` with a default temporary storage. This changes some tests like `test_no_init`.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

iycheng commented on python/ray/internal/storage.py:
```diff
@@ -152,6 +154,28 @@ def delete(self, path: str) -> bool:
         except FileNotFoundError:
             return False
 
+    def delete_prefix(self, path: str) -> bool:
```
We should rename it to delete_dir since this won't cover `delete /abc/def/prefix*`

suquark commented:
I doubt if s3 could work with ray internal storage. A lot of errors like `pyarrow.lib.ArrowInvalid: Unexpected query parameter in S3 URI: 'aws_session_token'`

ericl commented:
> I doubt if s3 could work with ray internal storage. A lot of errors like pyarrow.lib.ArrowInvalid: Unexpected query parameter in S3 URI: 'aws_session_token'. @iycheng any ideas about s3? I am not very familiar.

I haven't seen that before. Perhaps an upgrade of pyarrow or s3 library is needed here? In principle, we are following the same code path as Datasets for S3 access.

suquark commented:
@ericl @iycheng The CI failed for the same reason about S3. We can see the code tries to simulate accessing to AWS S3

https://github.com/ray-project/ray/blob/9f7170e4440b913438af0b7c80fd5d367115ff5d/python/ray/workflow/tests/conftest.py#L49-L60

However, even latest pyarrow does not support parameters like `endpoint_url` in S3. I wonder what should be the correct way of accessing cloud storage here?

ericl commented:
> So I put objects to S3 to the wrong path.

Is this just a bug in setting `self.root` for S3? I.e., it's set to `bucket/foo` instead of `/bucket/foo`, which means resolve() will treat it as a relative path?

suquark commented:
@ericl yes. `/bucket/foo` is an invalid S3 path, pyarrow would raise `E   pyarrow.lib.ArrowInvalid: Missing bucket name in S3 URI` with any S3 path starts with `/`. However, if you does not start with `/`, the path will be resolved to `<local_workdir>/<s3 path>`, which is also wrong.

I have fixed with issue in this PR.

There is another bug about S3: when `storage.get` a missing S3 key, it raises `OSError` instead of `FileNotFoundError`. `FileNotFoundError` in `storage.get` returns `None`, but `OSError` is unhandled. I also fixed this in the PR.

ericl commented:
Thanks @suquark . Is it worth splitting the fix to the storage API into a separate PR for clarity?

suquark commented:
@ericl The fix for storage is here: #24167

iycheng commented on python/ray/internal/storage.py:
```diff
@@ -129,6 +130,16 @@ def get(self, path: str) -> bytes:
         except FileNotFoundError:
             return None
+        except OSError as e:
+            # TODO(suquark): Instead of "FileNotFoundError", S3 filesystem raises
```
I'm ok with this now, but it seems should be something handled by pyarrow.

iycheng commented on python/ray/workflow/serialization.py:
```diff
@@ -113,53 +109,57 @@ async def export_stats(self) -> Dict[str, Any]:
 
-def obj_id_to_paths(workflow_id: str, object_id: str) -> List[str]:
-    return [workflow_id, OBJECTS_DIR, object_id]
+def _obj_id_to_key(object_id: str) -> str:
+    return os.path.join(OBJECTS_DIR, object_id)
```
I'm thinking about s3 and windows. Will this still work? I mean for s3, it's like 'a/b/c' and for windows, it's like 'a\b\c'

suquark replied:
it should still works. the storage would turn `/` into `\` with `.resolve()`

iycheng commented on python/ray/workflow/api.py:
```diff
@@ -39,45 +35,31 @@
+    _is_workflow_initialized = True
+
+
+def ensure_workflow_initialized() -> None:
+    if not _is_workflow_initialized:
+        init()
```
This shouldn't be a public API. Let's add `_` explicitly to ensure this.

iycheng commented on python/ray/workflow/tests/utils.py:
```diff
@@ -25,11 +21,7 @@ def check_global_mark(name="workflow"):
 
 
 def _alter_storage(new_storage):
```
should we just delete this method?

suquark replied:
yes, but I will leave it for the future to avoid linter errors. otherwise I have to cleanup those skipped tests

iycheng commented on python/ray/workflow/workflow_access.py:
```diff
@@ -393,26 +385,18 @@ def report_success(self, workflow_id: str) -> None:
+        ).remote()
         # No-op to ensure the actor is created before the driver exits.
-        ray.get(actor.get_storage_url.remote())
+        ray.get(actor.get_running_workflow.remote())
```
I feel a `ready` method is clearer.

iycheng reviewed (approved):
LGTM

aschuh-hf commented on python/ray/workflow/workflow_storage.py:
```diff
@@ -624,70 +608,51 @@ def get_latest_progress(self) -> "StepID":
+    def delete_workflow(self) -> None:
+        # TODO (Alex): There's a race condition here if someone tries to
+        # start the workflow between these ops.
+        found = self._storage.delete_dir("")
```
This line raises in error after upgrading to Ray 2.5.1:

```
  File "/opt/conda/envs/venv/lib/python3.10/site-packages/ray/workflow/workflow_storage.py", line 711, in delete_workflow
    found = self._storage.delete_dir("")
  File "pyarrow/_fs.pyx", line 473, in pyarrow._fs.FileSystem.delete_dir
  File "pyarrow/error.pxi", line 114, in pyarrow.lib.check_status
OSError: AWS Error [code 100]: Unable to parse ExceptionName: InvalidRequest Message: Missing required header for this request: Content-MD5 OR x-amz-checksum-*
```
